File: frontend/lib/brain.ts
import { api } from "./api";
import type { CellTranslation } from "./types";

export interface TranslatePromptConfig {
  prompt: string;
  provider_code: string | null;
  model: string | null;
  default_target_language: string;
}

export interface FixLinksPromptConfig {
  prompt: string;
  provider_code: string | null;
  model: string | null;
}

export interface GdocsMetaPromptConfig {
  // Prompt only — the Google-Docs import job picks its provider/model on the
  // upload modal, so there's nothing else to configure here.
  prompt: string;
}

export interface GdocsPairingPromptConfig {
  // Prompt only — maps each imported Doc to its Structure entry (slug source).
  prompt: string;
}

export interface BrainPrompts {
  translate: TranslatePromptConfig;
  fix_links: FixLinksPromptConfig;
  gdocs_meta: GdocsMetaPromptConfig;
  gdocs_pairing: GdocsPairingPromptConfig;
}

export function getBrainPrompts(): Promise<BrainPrompts> {
  return api<BrainPrompts>("/brain/prompts");
}

export function updateTranslateConfig(
  payload: TranslatePromptConfig,
): Promise<TranslatePromptConfig> {
  return api<TranslatePromptConfig>("/brain/prompts/translate", {
    method: "PUT",
    body: payload,
  });
}

export function updateFixLinksConfig(
  payload: FixLinksPromptConfig,
): Promise<FixLinksPromptConfig> {
  return api<FixLinksPromptConfig>("/brain/prompts/fix-links", {
    method: "PUT",
    body: payload,
  });
}

export function updateGdocsMetaConfig(
  payload: GdocsMetaPromptConfig,
): Promise<GdocsMetaPromptConfig> {
  return api<GdocsMetaPromptConfig>("/brain/prompts/gdocs-meta", {
    method: "PUT",
    body: payload,
  });
}

export function updateGdocsPairingConfig(
  payload: GdocsPairingPromptConfig,
): Promise<GdocsPairingPromptConfig> {
  return api<GdocsPairingPromptConfig>("/brain/prompts/gdocs-pairing", {
    method: "PUT",
    body: payload,
  });
}

export interface TranslateCellResponse extends CellTranslation {
  target_language: string;
  cached: boolean;
}

export function translateCell(
  tableId: number,
  rowId: number,
  columnId: number,
  targetLanguage: string,
  /** Set to true to bypass the server-side per-cell translation cache
   *  and force a fresh LLM call. Used by the Re-translate button. */
  force = false,
): Promise<TranslateCellResponse> {
  return api<TranslateCellResponse>(
    `/library/tables/${tableId}/cells/${rowId}/${columnId}/translate`,
    {
      method: "POST",
      body: { target_language: targetLanguage, force },
    },
  );
}

/** Translate a saved single-mode generation. Memoized server-side on
 *  ``generations.translations`` so opening the same saved generation
 *  later is free. */
export function translateGeneration(
  generationId: number,
  targetLanguage: string,
  force = false,
): Promise<TranslateCellResponse> {
  return api<TranslateCellResponse>(
    `/generations/${generationId}/translate`,
    {
      method: "POST",
      body: { target_language: targetLanguage, force },
    },
  );
}

/** Translate a prompt version's template content. Memoized server-side
 *  on ``prompt_versions.translations``; versions are immutable so the
 *  cache is permanent once written. */
export function translatePromptVersion(
  promptId: number,
  versionNumber: number,
  targetLanguage: string,
  force = false,
): Promise<TranslateCellResponse> {
  return api<TranslateCellResponse>(
    `/prompts/${promptId}/versions/${versionNumber}/translate`,
    {
      method: "POST",
      body: { target_language: targetLanguage, force },
    },
  );
}

/** Translate arbitrary text — used by ephemeral surfaces (test-prompt
 *  modal, unsaved single drafts). No memoization; every call hits the
 *  LLM. */
export function translateRawText(
  text: string,
  targetLanguage: string,
): Promise<TranslateCellResponse> {
  return api<TranslateCellResponse>("/brain/translate-text", {
    method: "POST",
    body: { text, target_language: targetLanguage },
  });
}

File: backend/app/api/brain.py
"""Brain — configurable system prompts for on-demand actions.

Today the only configurable surface is `translate`, used by the Translate
button in the bulk-table cell editor. The shape supports adding more
prompts (Summarize, Title-suggest, etc.) without another migration.

Endpoints:
  GET  /brain/prompts            — return all brain prompts (any role)
  PUT  /brain/prompts/translate  — admin only; update translate config

Cell-level translate runs live in ``library.py`` next to the rest of the
cell endpoints because they enforce the same table-ACL.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, require_role
from app.db.models import User
from app.db.session import get_db
from app.providers.base import ProviderError
from app.providers.registry import ProviderNotConfigured
from app.services.brain import (
    load_brain,
    now_iso,
    resolve_target_language,
    save_fix_links_config,
    save_gdocs_meta_config,
    save_gdocs_pairing_config,
    save_translate_config,
    translate_text,
)
from app.services.error_log import log_error
from app.services.usage import record_usage

router = APIRouter(
    prefix="/brain", tags=["brain"], dependencies=[Depends(get_current_user)]
)


class TranslatePromptRead(BaseModel):
    prompt: str
    provider_code: str | None = None
    model: str | None = None
    default_target_language: str = "ru"


class FixLinksPromptRead(BaseModel):
    prompt: str
    provider_code: str | None = None
    model: str | None = None


class GdocsMetaPromptRead(BaseModel):
    # Prompt only — the Google-Docs import job picks its provider/model on the
    # upload modal, so there is nothing to configure here besides the prompt.
    prompt: str


class GdocsPairingPromptRead(BaseModel):
    # Prompt only — maps each imported Doc to its Structure entry (slug source).
    prompt: str


class BrainPromptsRead(BaseModel):
    translate: TranslatePromptRead
    fix_links: FixLinksPromptRead
    gdocs_meta: GdocsMetaPromptRead
    gdocs_pairing: GdocsPairingPromptRead


class TranslatePromptUpdate(BaseModel):
    prompt: str = Field(min_length=1, max_length=10_000)
    provider_code: str | None = None
    model: str | None = None
    default_target_language: str = Field(
        default="ru", min_length=2, max_length=16
    )


class FixLinksPromptUpdate(BaseModel):
    prompt: str = Field(min_length=1, max_length=10_000)
    provider_code: str | None = None
    model: str | None = None


class GdocsMetaPromptUpdate(BaseModel):
    prompt: str = Field(min_length=1, max_length=10_000)


class GdocsPairingPromptUpdate(BaseModel):
    prompt: str = Field(min_length=1, max_length=10_000)


@router.get("/prompts", response_model=BrainPromptsRead)
async def get_prompts(db: AsyncSession = Depends(get_db)) -> BrainPromptsRead:
    cfg = await load_brain(db)
    return BrainPromptsRead.model_validate(cfg)


@router.put("/prompts/translate", response_model=TranslatePromptRead)
async def update_translate(
    payload: TranslatePromptUpdate,
    actor: User = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
) -> TranslatePromptRead:
    out = await save_translate_config(
        db,
        prompt=payload.prompt,
        provider_code=payload.provider_code,
        model=payload.model,
        default_target_language=payload.default_target_language,
        actor_id=actor.id,
    )
    return TranslatePromptRead.model_validate(out)


@router.put("/prompts/fix-links", response_model=FixLinksPromptRead)
async def update_fix_links(
    payload: FixLinksPromptUpdate,
    actor: User = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
) -> FixLinksPromptRead:
    out = await save_fix_links_config(
        db,
        prompt=payload.prompt,
        provider_code=payload.provider_code,
        model=payload.model,
        actor_id=actor.id,
    )
    return FixLinksPromptRead.model_validate(out)


@router.put("/prompts/gdocs-meta", response_model=GdocsMetaPromptRead)
async def update_gdocs_meta(
    payload: GdocsMetaPromptUpdate,
    actor: User = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
) -> GdocsMetaPromptRead:
    out = await save_gdocs_meta_config(
        db, prompt=payload.prompt, actor_id=actor.id
    )
    return GdocsMetaPromptRead.model_validate(out)


@router.put("/prompts/gdocs-pairing", response_model=GdocsPairingPromptRead)
async def update_gdocs_pairing(
    payload: GdocsPairingPromptUpdate,
    actor: User = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
) -> GdocsPairingPromptRead:
    out = await save_gdocs_pairing_config(
        db, prompt=payload.prompt, actor_id=actor.id
    )
    return GdocsPairingPromptRead.model_validate(out)


class TranslateTextRequest(BaseModel):
    """Stateless translate — used by surfaces that don't have a stable
    entity id to memoize against (test-modal results, unsaved single
    drafts). The text is sent inline; nothing is persisted."""

    text: str = Field(min_length=1, max_length=100_000)
    target_language: str | None = None


class TranslateTextResponse(BaseModel):
    text: str
    provider_used: str | None = None
    model_used: str | None = None
    translated_at: str
    target_language: str
    # Always false — the stateless endpoint never serves a cached entry.
    # Kept for shape parity with the memoized endpoints so the frontend
    # adapter doesn't have to branch.
    cached: bool = False


@router.post("/translate-text", response_model=TranslateTextResponse)
async def translate_raw_text(
    payload: TranslateTextRequest,
    actor: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TranslateTextResponse:
    """Translate arbitrary text using the brain prompt. Stateless —
    no memoization. Used by ephemeral surfaces (test-modal output,
    unsaved single draft) where there is no entity to cache against."""
    if not payload.text.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Nothing to translate.",
        )

    requested = await resolve_target_language(db, payload.target_language)

    try:
        text, code, model = await translate_text(
            db, source_text=payload.text, target_language=requested
        )
    except ProviderNotConfigured as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )
    except ProviderError as e:
        await log_error(
            db,
            source="api",
            category="provider_error",
            message=str(e),
            user_id=actor.id,
            status_code=getattr(e, "status_code", None),
            context={
                "endpoint": "/brain/translate-text",
                "target_language": requested,
                "text_len": len(payload.text),
            },
        )
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY, detail=str(e)
        )

    await record_usage(
        db,
        user_id=actor.id,
        provider_code=code,
        model=model,
        prompt_tokens=None,
        completion_tokens=None,
        source="brain_translate",
        source_ref={"target_language": requested, "kind": "raw_text"},
    )

    return TranslateTextResponse(
        text=text,
        provider_used=code,
        model_used=model,
        translated_at=now_iso(),
        target_language=requested,
        cached=False,
    )
